drv/gimlet-seq-server/src/main.rs

```diff
@@ -633,6 +680,34 @@ impl<S: SpiServer> NotificationHandler for ServerImpl<S> {
     }
 }
 
+fn retry_i2c_txn<T, E>(
+    mut txn: impl FnMut() -> Result<T, E>,
+) -> Result<T, i2c::ResponseCode>
+where
+    i2c::ResponseCode: From<E>,
+{
+    // Chosen by fair dice roll, seems reasonable-ish?
+    let mut retries_remaining = 3;
+    loop {
```

Comment:
I might consider structuring this as a `for _ in 0..retries_remaining` just to avoid accidentally creating an infinite loop through a logic error below.

Reply:
that was what i was going to do. but, we want to return the last error when we run out of retries, which means that we need to keep it around when we fall through to the end of the loop, which felt a bit ugly; it would have to be something like this:
```rust
// Chosen arbitrarily, but don't worry! We'll never actually return this --- it's 
// just because we need to initialize this to *something*.
let mut last_error = i2c::ResponseCode::BadResponse;
for retries_remaining in (0..3).rev() {
        match txn() {
            Ok(x) => return Ok(x),
            Err(e) => {
                let code = e.into();
                ringbuf_entry!(Trace::I2cFault {
                    retries_remaining,
                    code,
                });
                last_error = code;
            }
}

Err(code)
```

in particular, i felt like having to assign `last_error` to _something_, even though we've never actually seen that error, felt a bit unfortunate.

what do you think?

Reply:
Hmmmm. Fair point. Current structure's fine.

Reply:
A third option, which is _also_ not my favorite, is to do something like this:
```rust
    for retries_remaining in (0..MAX_RETRIES).rev() {
        match txn() {
            Ok(x) => return Ok(x),
            Err(error) => {
                let code = error.into();
                ringbuf_entry!(Trace::I2cFault {
                    retries_remaining,
                    code,
                });
                if retries_remaining == 0 {
                    return Err(code);
                }
            }
        }
    }

    unreachable!()
```

Now, we've gotten rid of the arbitrary error code that we'll never actually return, *and* the infinite loop that isn't actually infinite, but this comes at the cost of an `unreachable!()`, and having to write `unreachable!()` also makes me feel kind of sad. 

If you have a preference for any of these bad options, I'll happily go with that one --- I don't really have strong feelings about which is the least objectionable. :)